### PR TITLE
Default mail template

### LINKED ISSUE #82: Add a default email template
This will make it easier to add nicely styled emails.

### PR DESCRIPTION
**Pull request by tijsverkoyen**
# Sending mails

Sending mails is a trivial part of an application, but is something that in 
most cases costing a big amount of time.

We use the standard way of sending mails in Symfony, which uses SwiftMailer.

In addition a `MessageFactory` has been added, which will enable you to create 
messages that can be send with the `mailer`-service.

There a some public methods you can use:

* `createHtmlMessage` which enables you to create an message that will contain HTML
    If you don't provide a plainText alternative it will be generated from the provided HTML.
    The content will be wrapped in the default template.
    
* `createPlainTextMessage`, which enables you to send just plain-text emails.

There are some other public methods which are exposed and can help you:

* `wrapInTemplate`, which will wrap the provided content in a nice looking mail-template.

## Basic example

```php
// ...
/** @var /SumoCoders\FrameworkCoreBundle\Mail\MessageFactory/MessageFactory $messageFactory */
$messageFactory = $this->get('framework.message_factory');

// create a simple message
$message = $messageFactory->createHtmlMessage(
    'the subject',
    '<p>foo bar</p>'
);

// set some extra properties, just like you would do with a normal \Swift_Message
$message->setTo(
    $this->getParameter('mailer_default_to_email')
);

// send it
$this->get('mailer')->send($message);
```

This will fix #82 
If this is merged I will add an example into the ExampleBundle

### CODE REVIEW

**Comment by tijsverkoyen** [comment]
@sumocoders/php-team, @mathiashelin, @Katrienvh any feedback would be appreciated.

File: app/config/parameters.yml.dist [hunk 1]
```diff
@@ -14,6 +14,15 @@ parameters:
...
     mailer_transport: mail
+    mailer_host: mail.hostbots.be
+    mailer_user:
+    mailer_password:
+    mailer_default_sender_name:
+    mailer_default_sender_email:
+    mailer_default_to_name:
+    mailer_default_to_email:
+    mailer_default_reply_to_name: "%mailer_default_sender_name%"
+    mailer_default_reply_to_email:
```
Comment (WouterSioen) [inline app/config/parameters.yml.dist]: I'd add the default_sender_email as default here

File: composer.json [hunk 2]
```diff
@@ -32,7 +32,8 @@
         "sumocoders/framework-error-bundle": "~2.0",
         "sumocoders/framework-user-bundle": "~2.0",
         "sumocoders/framework-search-bundle": "~2.0",
-        "sumocoders/framework-example-bundle": "^3.0"
+        "sumocoders/framework-example-bundle": "^3.0",
+        "tijsverkoyen/css-to-inline-styles": "^1.5@dev"
```
Comment (WouterSioen) [inline composer.json]: Is it a good idea to add a dev dependency? I'd prefer a stable version.

File: src/SumoCoders/FrameworkCoreBundle/Mail/MessageFactory.php [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+
+namespace SumoCoders\FrameworkCoreBundle\Mail;
+
+use Symfony\Bundle\FrameworkBundle\Templating\EngineInterface;
+use TijsVerkoyen\CssToInlineStyles\CssToInlineStyles;
+
+class MessageFactory
+{
+    /**
+     * @var array
```
Comment (WouterSioen) [inline src/SumoCoders/FrameworkCoreBundle/Mail/MessageFactory.php]: this should be `array|string` because you use a string when no name is given

File: composer.json (hunk 2 above)
Comment (tijsverkoyen) [inline composer.json]: :+1:

File: src/SumoCoders/FrameworkCoreBundle/Mail/MessageFactory.php [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+        $css = file_get_contents($this->cssPath);
+        $html = $this->template->render(
+            $this->templatePath,
+            array(
+                'content' => $content,
+                'css' => $css,
+            )
+        );
+
+        $cssToInlineStyles = new CssToInlineStyles(
```
Comment (WouterSioen) [inline src/SumoCoders/FrameworkCoreBundle/Mail/MessageFactory.php]: It would be cleaner if this is a service that gets injected in the constructor, but I guess you can't construct it without html and css?

File: composer.json (hunk 2 above)
Comment (tijsverkoyen) [inline composer.json]: I will release a new version soon, and i will change it here.

File: src/SumoCoders/FrameworkCoreBundle/Mail/MessageFactory.php [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+
+    /**
+     * Create a message
+     *
+     * @param string|null $subject
+     * @param string|null $html
+     * @param string|null $alternative
+     * @return \Swift_Message
+     */
+    protected function createMessage($subject = null, $html = null, $alternative = null)
```
Comment (WouterSioen) [inline src/SumoCoders/FrameworkCoreBundle/Mail/MessageFactory.php]: Why would you allow creating messages without subject? Emails without subject should die!

File: src/SumoCoders/FrameworkCoreBundle/Tests/Mail/MessageFactoryTest.php [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+     */
+    protected function tearDown()
+    {
+        $this->messageFactory = null;
+    }
+
+    /**
+     * @return \PHPUnit_Framework_MockObject_MockObject
+     */
+    protected function getTemplating()
```
Comment (WouterSioen) [inline src/SumoCoders/FrameworkCoreBundle/Tests/Mail/MessageFactoryTest.php]: I'd call this "getTemplateingMock".

But why don't you use a real Twig instance here? This would enable you to test the creation of html emails. Your html test now does not test anything about the rendering.

File: src/SumoCoders/FrameworkCoreBundle/Mail/MessageFactory.php (hunk 4 above)
Comment (tijsverkoyen) [inline src/SumoCoders/FrameworkCoreBundle/Mail/MessageFactory.php]: I know, but that should be changed in the original class

File: src/SumoCoders/FrameworkCoreBundle/Mail/MessageFactory.php [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+
+use Symfony\Bundle\FrameworkBundle\Templating\EngineInterface;
+use TijsVerkoyen\CssToInlineStyles\CssToInlineStyles;
+
+class MessageFactory
+{
+    /**
+     * @var array
+     */
+    protected $sender = array();
```
Comment (WouterSioen) [inline src/SumoCoders/FrameworkCoreBundle/Mail/MessageFactory.php]: why everything protected? Do you want this class to be extendable?

File: src/SumoCoders/FrameworkCoreBundle/Mail/MessageFactory.php (hunk 5 above)
Comment (tijsverkoyen) [inline src/SumoCoders/FrameworkCoreBundle/Mail/MessageFactory.php]: I want it to be as close to the `\Swift_Message` as possible.

File: composer.json (hunk 2 above)
Comment (WouterSioen) [inline composer.json]: :+1:

File: src/SumoCoders/FrameworkCoreBundle/Mail/MessageFactory.php [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+            $message->setSubject($subject);
+        }
+
+        if ($html != '' && $alternative != '') {
+            $message->setBody(
+                $this->wrapInTemplate($html),
+                'text/html'
+            );
+            $message->addPart($alternative, 'text/plain');
+        } elseif ($html != '' && $alternative == '') {
```
Comment (WouterSioen) [inline src/SumoCoders/FrameworkCoreBundle/Mail/MessageFactory.php]: I think this could be a lot cleaner if you used early returns instead of elseif statements.

File: src/SumoCoders/FrameworkCoreBundle/Mail/MessageFactory.php (hunk 7 above)
Comment (tijsverkoyen) [inline src/SumoCoders/FrameworkCoreBundle/Mail/MessageFactory.php]: why should it not be extendable?

File: src/SumoCoders/FrameworkCoreBundle/Mail/MessageFactory.php [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+        );
+
+        $cssToInlineStyles = new CssToInlineStyles(
+            $html,
+            $css
+        );
+
+        $content = $cssToInlineStyles->convert();
+
+        return $content;
```
Comment (WouterSioen) [inline src/SumoCoders/FrameworkCoreBundle/Mail/MessageFactory.php]: no need to create this content var, you can directly return from the convert call;

File: src/SumoCoders/FrameworkCoreBundle/Tests/Mail/MessageFactoryTest.php [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+    {
+        $templating = $this->getMock('\Symfony\Bundle\FrameworkBundle\Templating\EngineInterface');
+        $templating
+            ->method('render')
+            ->willReturn('<html><head></head><body><p>And I, le content</p></body></html>');
+
+        return $templating;
+    }
+
+    public function testIfTheDefaultsAreEmptyWhenNotSet()
```
Comment (WouterSioen) [inline src/SumoCoders/FrameworkCoreBundle/Tests/Mail/MessageFactoryTest.php]: testIf is kind of an anti pattern in test naming. You should be able to read the test name as a sentence if you strip 'test' from it. But "If the defaults are empty when nog set" is not a great sentence.

File: src/SumoCoders/FrameworkCoreBundle/Tests/Mail/MessageFactoryTest.php [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+    {
+        /** @var \Swift_Message $message */
```
Comment (WouterSioen) [inline src/SumoCoders/FrameworkCoreBundle/Tests/Mail/MessageFactoryTest.php]: why is this needed? The @return statement is correct in the messageFactory, so this is just polluting the code I guess?

File: src/SumoCoders/FrameworkCoreBundle/Tests/Mail/MessageFactoryTest.php [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+        $this->assertEmpty($message->getTo());
+    }
+
+    public function testIfTheDefaultsAreCorrectWhenSet()
+    {
+        $this->messageFactory->setDefaultSender('from@example.com', 'John From');
+        $this->messageFactory->setDefaultReplyTo('reply-to@example.com', 'John Reply To');
+        $this->messageFactory->setDefaultTo('to@example.com', 'John To');
+
+        /** @var \Swift_Message $message */
```
Comment (WouterSioen) [inline src/SumoCoders/FrameworkCoreBundle/Tests/Mail/MessageFactoryTest.php]: idem

File: src/SumoCoders/FrameworkCoreBundle/Tests/Mail/MessageFactoryTest.php [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+        $this->assertEquals(array('reply-to@example.com' => 'John Reply To'), $message->getReplyTo());
+        $this->assertEquals(array('to@example.com' => 'John To'), $message->getTo());
+    }
+
+    public function testCreationOfPlainTextMessage()
+    {
+        $subject = 'I am le subject';
+        $content = 'And I, le content';
+
+        /** @var \Swift_Message $message */
```
Comment (WouterSioen) [inline src/SumoCoders/FrameworkCoreBundle/Tests/Mail/MessageFactoryTest.php]: idem

File: src/SumoCoders/FrameworkCoreBundle/Tests/Mail/MessageFactoryTest.php [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+    {
+        $subject = 'I am le subject';
+        $content = '<p>And I, le content</p>';
+        $expectedResult = <<<EOF
+<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN" "http://www.w3.org/TR/REC-html40/loose.dtd">
+<html><head></head><body><p>And I, le content</p></body></html>
+
+EOF;
+
+        /** @var \Swift_Message $message */
```
Comment (WouterSioen) [inline src/SumoCoders/FrameworkCoreBundle/Tests/Mail/MessageFactoryTest.php]: idem

File: src/SumoCoders/FrameworkCoreBundle/Mail/MessageFactory.php (hunk 7 above)
Comment (WouterSioen) [inline src/SumoCoders/FrameworkCoreBundle/Mail/MessageFactory.php]: Why should it? The open closed principle states that "software entities (classes, modules, functions, etc.) should be open for extension, but closed for modification".

You now open your class for modification, so everyone can extend it and start to do stupid things with it.

I think it's a better default to make classes not extendable. If you want to extend them, it's still possible by changing the properties you want to extend. IMO, making properties private and classes final is a sane default in most cases.

File: src/SumoCoders/FrameworkCoreBundle/Mail/MessageFactory.php (hunk 5 above)
Comment (WouterSioen) [inline src/SumoCoders/FrameworkCoreBundle/Mail/MessageFactory.php]: Ok, I agree

File: src/SumoCoders/FrameworkCoreBundle/Mail/MessageFactory.php (hunk 4 above)
Comment (WouterSioen) [inline src/SumoCoders/FrameworkCoreBundle/Mail/MessageFactory.php]: hmm, that would break backwards compatibility, so let's keep it this way for now :)

File: src/SumoCoders/FrameworkCoreBundle/Tests/Mail/MessageFactoryTest.php (hunk 6 above)
Comment (tijsverkoyen) [inline src/SumoCoders/FrameworkCoreBundle/Tests/Mail/MessageFactoryTest.php]: fixed in d4a9e3758b5e86ab6ddabff340687f8385c90b1f

File: src/SumoCoders/FrameworkCoreBundle/Mail/MessageFactory.php (hunk 7 above)
Comment (tijsverkoyen) [inline src/SumoCoders/FrameworkCoreBundle/Mail/MessageFactory.php]: fixed in d4a9e3758b5e86ab6ddabff340687f8385c90b1f

File: src/SumoCoders/FrameworkCoreBundle/Mail/MessageFactory.php (hunk 8 above)
Comment (tijsverkoyen) [inline src/SumoCoders/FrameworkCoreBundle/Mail/MessageFactory.php]: fixed in d4a9e3758b5e86ab6ddabff340687f8385c90b1f

File: src/SumoCoders/FrameworkCoreBundle/Mail/MessageFactory.php (hunk 9 above)
Comment (tijsverkoyen) [inline src/SumoCoders/FrameworkCoreBundle/Mail/MessageFactory.php]: fixed in d4a9e3758b5e86ab6ddabff340687f8385c90b1f

File: src/SumoCoders/FrameworkCoreBundle/Tests/Mail/MessageFactoryTest.php (hunk 10 above)
Comment (tijsverkoyen) [inline src/SumoCoders/FrameworkCoreBundle/Tests/Mail/MessageFactoryTest.php]: fixed in d4a9e3758b5e86ab6ddabff340687f8385c90b1f

File: src/SumoCoders/FrameworkCoreBundle/Tests/Mail/MessageFactoryTest.php (hunk 11 above)
Comment (tijsverkoyen) [inline src/SumoCoders/FrameworkCoreBundle/Tests/Mail/MessageFactoryTest.php]: fixed in d4a9e3758b5e86ab6ddabff340687f8385c90b1f, was used while I was debugging stuff.

File: src/SumoCoders/FrameworkCoreBundle/Tests/Mail/MessageFactoryTest.php (hunk 12 above)
Comment (tijsverkoyen) [inline src/SumoCoders/FrameworkCoreBundle/Tests/Mail/MessageFactoryTest.php]: fixed in d4a9e3758b5e86ab6ddabff340687f8385c90b1f

File: src/SumoCoders/FrameworkCoreBundle/Tests/Mail/MessageFactoryTest.php (hunk 13 above)
Comment (tijsverkoyen) [inline src/SumoCoders/FrameworkCoreBundle/Tests/Mail/MessageFactoryTest.php]: fixed in d4a9e3758b5e86ab6ddabff340687f8385c90b1f

File: src/SumoCoders/FrameworkCoreBundle/Tests/Mail/MessageFactoryTest.php (hunk 14 above)
Comment (tijsverkoyen) [inline src/SumoCoders/FrameworkCoreBundle/Tests/Mail/MessageFactoryTest.php]: fixed in d4a9e3758b5e86ab6ddabff340687f8385c90b1f

File: src/SumoCoders/FrameworkCoreBundle/Mail/MessageFactory.php [hunk 15]
```diff
@@ -0,0 +1,249 @@
...
+    public function createDefaultMessage()
+    {
+        $message = \Swift_Message::newInstance();
+
+        if (!empty($this->sender)) {
+            $message->setFrom($this->sender);
+        }
+
+        if (!empty($this->replyTo)) {
+            $message->setReplyTo($this->replyTo);
```
Comment (WouterSioen) [inline src/SumoCoders/FrameworkCoreBundle/Mail/MessageFactory.php]: Scrutinizer zaagt hiervan:

> It seems like $this->replyTo can also be of type array; however, Swift_Mime_SimpleMessage::setReplyTo() does only seem to accept string, maybe add an additional type check?

Is scrutinizer hier mis?

File: src/SumoCoders/FrameworkCoreBundle/Mail/MessageFactory.php [hunk 16]
```diff
@@ -0,0 +1,249 @@
...
+
+    /**
+     * @var string
+     */
+    private $cssPath;
+
+    /**
+     * MessageFactory constructor.
+     *
+     * @param EngineInterface $template
```
Comment (WouterSioen) [inline src/SumoCoders/FrameworkCoreBundle/Mail/MessageFactory.php]: phpdoc is incomplete

File: src/SumoCoders/FrameworkCoreBundle/Mail/MessageFactory.php (hunk 15 above)
Comment (tijsverkoyen) [inline src/SumoCoders/FrameworkCoreBundle/Mail/MessageFactory.php]: Scrutiniser is juist, het is de documentatie van `\Swift_Mime_SimpleMessage` die fout is.

Comment (WouterSioen) [inline src/SumoCoders/FrameworkCoreBundle/Mail/MessageFactory.php]: ok

File: src/SumoCoders/FrameworkCoreBundle/Mail/MessageFactory.php (hunk 16 above)
Comment (tijsverkoyen) [inline src/SumoCoders/FrameworkCoreBundle/Mail/MessageFactory.php]: fixed in d2bd7162b356461dbc84dca75dbb857638cc6154
